File: Medium/424_Longest_Repeating_Character_Replacement/Longest_Repeating_Character_Replacement.go
package longestrepeatingcharacterreplacement

func characterReplacement(s string, k int) int {
	countMap := make(map[byte]int)
	answer := 0
	maxFreq := 0
	rightIndex := 0

	// for leftIndex := 0; leftIndex < len(s); leftIndex++ {
	for leftIndex := range s {
		for rightIndex < len(s) && rightIndex-leftIndex-maxFreq <= k {
			countMap[s[rightIndex]] += 1
			maxFreq = max(maxFreq, countMap[s[rightIndex]])
			rightIndex += 1
		}

		// Update the length of the longest substring
		if rightIndex-leftIndex-maxFreq > k {
			// The new character is different from the current maxFreq character
			answer = max(answer, rightIndex-leftIndex-1)
		} else {
			// rightIndex has reached the rightmost position
			answer = max(answer, rightIndex-leftIndex)
		}

		countMap[s[leftIndex]] -= 1

	}

	return answer

}

File: Medium/424_Longest_Repeating_Character_Replacement/longest_repeating_character_replacement.py
from collections import defaultdict

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count_map = defaultdict(int)
        answer = 0
        max_freq = 0
        right = 0

        for left in range(len(s)):
            while right < len(s) and (right - left - max_freq <= k) :
                count_map[s[right]] += 1
                max_freq = max(max_freq, count_map[s[right]])
                right += 1


            if right - left - max_freq > k:
                answer = max(answer, right - left - 1)
            else:
                answer = max(answer, right - left)

            count_map[s[left]] -= 1

        return answer
